src/Goteo/Console/Command/OpenDataCommand.php

```diff
@@ -223,6 +299,20 @@ private function extractCallInvestsData(Call $call): void {
 
         if ( $file->upload('/tmp/' . $fileName, $fileName) ) {
             $this->log("\nUpload of file {$fileName} completed!", [], 'info');
+            $dataSetRepository = new DataSetRepository();
+            try {
+                $dataSet = $dataSetRepository->getLastByCallAndType([$call->id], DataSet::DATASET_TYPE_INVESTS);
+                $dataSet->setUrl($file->get_path() . $fileName);
+                $dataSetRepository->persist($dataSet);
+                $call->addDataSet($dataSet, 1);
+            } catch (ModelNotFoundException $e) {
+                $dataSet = new DataSet();
+                $dataSet->setTitle("$call->name");
+                $dataSet->setUrl($file->get_path() . $fileName);
+                $dataSet->setType(DataSet::DATASET_TYPE_INVESTS);
+                $dataSetRepository->persist($dataSet);
+                $call->addDataSet($dataSet, 1);
+            }
```

Comment:
@davidbeig the `try` and `catch` blocks of these lines could be refactored to avoid code duplication. Specially the `catch` block, which is identical in all cases except for the argument passed to the `setType()` method.

We could have a method that creates the `DataSet`, with an `int $type` parameter, and reuse it later in all `catch` blocks of this `Command`.

The `try` block is still likely to have a refactor, with a little bit more of work, but definitely doable. This will save us a tone of lines of code and lots of work in a future refactor. :wink: